### DISCUSSION TITLE
Laravel sanctum:Method Illuminate\Auth\RequestGuard::viaRemember does not exist

### DISCUSSION THREAD

**Discussion by findridoy** [laravel/framework#36423] [Q&A]
I am having problem with laravel sanctum.

I am using laravel sanctum with my vue js frontend which is eventually translating to SPA. Which haven't done yet. 
I am getting this error on using auth:sanctum middleware.

BadMethodCallException
Method Illuminate\Auth\RequestGuard::viaRemember does not exist.

**Comment by ryantxr** [comment]
I have the same problem.

**Reply by hamidsadeghi431** [reply]
Hi I have the same problem too

**Comment by fritz-c** [comment]
I've seen [a couple valid-looking solutions in this issue](https://github.com/tymondesigns/jwt-auth/issues/1436), where the presence of session authentication middleware was loading a session that then caused the sanctum auth to fail.

My problem, however, only happens during tests, and after extensive digging it appears that it is caused by the approach Laravel takes to unit test setup, where the Laravel application is only booted once at the beginning of each test, and not per-request, as would normally happen. Since it does not get torn down after each request, calls like `app()->scoped(StatefulGuard::class, function() { /*...*/ })` inside app code get persisted between requests.

In my case, Laravel Nova had a call of `$app->scoped(StatefulGuard::class, static fn () => Auth::guard(Util::userGuard()));` inside its code. When I made a request to a Nova url followed by a request to the Jetstream-side of my app, the second request used the old, Nova-established container binding to get the `StatefulGuard` instance, rather than the binding provided by `Laravel\Fortify\FortifyServiceProvider`. Due to [some logic in Laravel's Authenticate middleware](https://github.com/laravel/framework/blob/3faeb0726cb55b4f672bf549cc93b716545b84c2/src/Illuminate/Auth/Middleware/Authenticate.php#L83), the config value stored in `config('auth.defaults.guard')` is overwritten partway through the request to use Sanctum if you use [Jetstream's default middleware stack for your routes](https://github.com/laravel/jetstream/blob/0ce161d3472905e3253dace2eab77290982ce591/src/Console/InstallCommand.php#L326-L330), which has `auth:sanctum` whether you use stateless auth or not. The combination of the stale service container binding and the auth default getting changed to sanctum resulted in the test request attempting to use a session-less guard in a request that already had a session value set to it, which resulted in the session-less guard being treated as if it had session-related methods on it, resulting in the error here.

My fix for my particular case was to change `auth:sanctum` in my route definitions to `auth`, as Sanctum is not necessary for my use case.

There may be room for a more general fix in Laravel's test runner logic, but I'm not sure exactly what you could do without breaking other functionality at the same time.
